go/vt/schemadiff/schema.go

```diff
@@ -282,3 +285,150 @@ func (s *Schema) Diff(other *Schema, hints *DiffHints) (diffs []EntityDiff, err
 	}
 	return diffs, err
 }
+
+// Entity returns an entity by name, or nil if nonexistent
+func (s *Schema) Entity(name string) Entity {
+	return s.named[name]
+}
+
+// ToStatements returns an ordered list of statements which can be applied to create the schema
+func (s *Schema) ToStatements() []sqlparser.Statement {
+	stmts := []sqlparser.Statement{}
+	for _, e := range s.Entities() {
+		stmts = append(stmts, e.Create().Statement())
+	}
+	return stmts
+}
+
+// ToQueries returns an ordered list of queries which can be applied to create the schema
+func (s *Schema) ToQueries() []string {
+	queries := []string{}
+	for _, e := range s.Entities() {
+		queries = append(queries, e.Create().CanonicalStatementString())
+	}
+	return queries
+}
+
+// ToSQL returns a SQL blob with ordered sequence of queries which can be applied to create the schema
+func (s *Schema) ToSQL() string {
+	var buf bytes.Buffer
+	for _, query := range s.ToQueries() {
+		buf.WriteString(query)
+		buf.WriteString(";\n")
+	}
+	return buf.String()
+}
+
+// apply attempts to apply given list of diffs to this object.
+// These diffs are CREATE/DROP/ALTER TABLE/VIEW.
+func (s *Schema) apply(diffs []EntityDiff) error {
+	for _, diff := range diffs {
+		switch diff := diff.(type) {
+		case *CreateTableEntityDiff:
+			// We expect the table to not exist
+			if _, ok := s.named[diff.createTable.Table.Name.String()]; ok {
+				return ErrApplyDuplicateTableOrView
+			}
+			s.tables = append(s.tables, &CreateTableEntity{CreateTable: *diff.createTable})
+		case *CreateViewEntityDiff:
+			// We expect the view to not exist
+			if _, ok := s.named[diff.createView.ViewName.Name.String()]; ok {
+				return ErrApplyDuplicateTableOrView
+			}
+			s.views = append(s.views, &CreateViewEntity{CreateView: *diff.createView})
+		case *DropTableEntityDiff:
+			// We expect the table to exist
+			found := false
+			for i, t := range s.tables {
+				if t.Table.Name.String() == diff.from.Table.Name.String() {
+					s.tables = append(s.tables[0:i], s.tables[i+1:]...)
+					delete(s.named, t.Table.Name.String())
+					found = true
+					break
+				}
+			}
+			if !found {
+				return ErrApplyTableNotFound
+			}
+		case *DropViewEntityDiff:
+			// We expect the view to exist
+			found := false
+			for i, v := range s.views {
+				if v.ViewName.Name.String() == diff.from.ViewName.Name.String() {
+					s.views = append(s.views[0:i], s.views[i+1:]...)
+					delete(s.named, v.ViewName.Name.String())
+					found = true
+					break
+				}
+			}
+			if !found {
+				return ErrApplyViewNotFound
+			}
+		case *AlterTableEntityDiff:
+			// We expect the table to exist
+			found := false
+			for i, t := range s.tables {
+				if t.Table.Name.String() == diff.from.Table.Name.String() {
+					to, err := t.Apply(diff)
+					if err != nil {
+						return err
+					}
+					toCreateTableEntity, ok := to.(*CreateTableEntity)
+					if !ok {
+						return ErrEntityTypeMismatch
+					}
+					s.tables[i] = toCreateTableEntity
```

Comment:
Does `s.named` need to be updated here?

Reply:
It does not, because it is implied that the view name already exists in `named`, and an update would just re-apply same name.

Reply:
@shlomi-noach I'm probably missing something, but doesn't then calling for example `Entity(name)` return the old entity without the change applied? Or is this updated then somewhere else?

The reason I was asking is because this does get updated on create / drop so that's what caught my eye.

Reply:
@dbussink gotcha. `named` is only used in _this_ function to determine existence/nonexistence of an entity. At the end of this function, we call `normalize()` which completely rebuilds `named`. We could update the entity value here, but it then gets deleted and recreated in `normalize` anyhow.

Reply:
@shlomi-noach Ah ok, if that's the case I think there's an inconsistency though in the add / drop logic. 

When a create is done, we don't add it to `named` so if the `diffs` slice has multiple creates for the same table, it would not error but have a last one wins.

On the other hand, if `diffs` contains multiple drops for the same table, it would error on the second one since it does delete from `named`. 

I have no super strong opinion on what the behavior should be, but I think it should be consistent between create and drop?

Reply:
You're right, there is an inconsistency. Fixing.

Reply:
Please see updates to `Schema.apply()` and consistent use of `s.named`